### ISSUE TITLE
Allow for Some Kind of Authentication to prevent Ballot Stuffing

### ISSUE THREAD

**Issue by mudlock** [electology/approval_frame#7]
Currently, nothing limits anyone from voting as many times as they want on anything.

Restrict that. (Configurable by poll.) Some possibilities (implement any or all):

Just a friendly cookie: "Hey, you already voted on this."
User accounts (verify by email?)
reCaptcha
OpenAuth
Question ids just increment. How about adding a secret key to get in (equiv. to gDocs "anyone who knows the url"?)

**Comment by ashimaathri** [comment]
I think this is really important for our application because currently nothing stops the user from voting multiple times and messing up the results. 
However, we need to keep in mind that the iframe would most probably be embedded in an external page that we have no control over. So, should we enable authentication with a well known provider like Google/Facebook or should we have a local authentication system which authenticates the user against the approval voting app itself using the registration app?
I see some support for [using the IP](https://github.com/electology/approval_frame/blob/master/approval_polls/views.py#L42) to uniquely identify a user, but I don't think that works yet.

**Comment by ashimaathri** [comment]
Comments from Aaron

* Limit poll submissions to one per e-mail.
* To submit vote, must have e-mail.
* "Vote again" should be "change vote".

**Comment by adihemdev** [comment]
I found that Django already has a way to implement the second requirement.  This is from the Django docs. . I'd like to go ahead and tackle this using the steps provided in the docs.

**Comment by adihemdev** [comment]
"To submit vote, must have email".......Is this the same as "to submit vote, must be logged in". In that case, a @login_required should suffice.

**Comment by adihemdev** [comment]
Ok..this is my understanding so far:

If only one submission is allowed per email, then Django's user class must be extended to include a "vote" field, like so?
vote=models.ForeignKey(Vote)

so that a vote can have multiple users associated with it, but not the other way around

To summarize, what I'm saying is, why not create a custom user class, which inherits from django.contrib.auth.models.User, like this
class CustomUser(User):


and add the vote field as described above inside it





The second part, (user must have email to submit vote), can be taken care of this way:
if(request.user.email):     //this calls the getter method 
   THEN go to the page that shows the results
else:
    please enter a valid email









Please correct me if I'm wrong

**Comment by ashimaathri** [comment]
Thanks for looking at this issue @adihemdev! Yes, I agree that we should use Django's built-in user authentication system as we're already using it for creating polls. As you suggested, using the decorator @login_required will satisfy requirement 2.

I think the Vote and Ballot tables are a little confusing at first glance. The ballot is what the user casts and in our case we want to make sure that the user can do so only once. The vote is actually the choice that the user has selected for a particular poll, and in our case (approval voting), the user can select multiple options so he/she can have multiple votes (entries in the Vote table). 

One way to achieve this is as follows:

Make the user a foreign key for the Ballot table like so
```
from Django.contrib.auth import models.User
...
class Ballot(models.Model):
  ...
  user = models.ForeignKey(models.User)
  ...
```

This again ensures requirement 2 (a ballot is always associated with a user - foreign key constraint).

In the [vote method](https://github.com/electology/approval_frame/blob/master/approval_polls/views.py#L36) of the view, before creating the ballot, check if there already exists a ballot cast by a particular user for a particular poll (query the db). If yes, then notify the user that they've already cast their vote. If not, then we continue as before (create the ballot and redirect).

The Vote table can not be a foreign key for the User table because then a user can only vote once and never again (not once per poll, but once globally). It also means that a user must be associated with a vote on signup (foreign key must be present), which is not possible.

**Comment by adihemdev** [comment]
Ashima,
I got confused with the definition of vote vs Ballot. Thank you for clarifying that! I am a little new to GitHub, so before I modify anything, I just want to make sure I am clear with the process

I was looking online and I found this link


https://guides.github.com/activities/hello-world/

It looks like a quick way to get yourself familiar with GIT. I recommend anyone else who is new to try it out

Aditya

**Comment by ashimaathri** [comment]
@adihemdev That looks like an excellent resource. Could you put it up on the forum so that everyone can see it? Thanks!

**Comment by Daniel-Avila** [comment]
Is someone already working on this? It seems like a very simple fix. I would hate to duplicate work efforts. 

It is really a simple matter of adding the user as a foreign key on the ballot and checking against that. 

Can I take this one if it is not actively being worked on?

**Comment by rupalkhilari** [comment]
Hi @Daniel-Avila , I had begun working on https://github.com/electology/approval_frame/issues/16 which I believe is a subset of this issue (the part that you mentioned) - there seemed to be an overlap here.  I realized that I will be quite busy until the end of this month, so please feel free to look into that if you think it could be completed earlier.

**Comment by rupalkhilari** [comment]
Have added a 'Poll Settings' panel which allows allows the user to choose whether he/she prefers only registered users to vote/ or anyone to vote. This probably partially addresses the issue of ballot stuffing. 
https://github.com/electology/approval_frame/issues/16

**Comment by rupalkhilari** [comment]
Since the poll now has an option of allowing only registered users to vote #16 , with the integration of social authentication https://github.com/electology/approval_frame/issues/69 and with allowing users with the specified email address to vote https://github.com/electology/approval_frame/issues/47 I think we can close this for now. We might investigate cookies at a later stage for which we can raise a separate ticket.
